src/NLog/Logger.cs

```diff
@@ -557,58 +612,94 @@ internal void Initialize(string name, LoggerConfiguration loggerConfiguration, L
+            var targetsForLevel = GetTargetsForLevel(level);
+            if (targetsForLevel != null)
+            {
+                var logEvent = LogEventInfo.Create(level, Name, ex, Factory.DefaultCultureInfo, message, args);
+                WriteToTargets(logEvent, targetsForLevel);
+            }
         }
 
-        private void WriteToTargets(LogEventInfo logEvent)
+        private void WriteToTargets(LogLevel level, Exception ex, IFormatProvider formatProvider, [Localizable(false)] string message, object[] args)
         {
-            LoggerImpl.Write(_loggerType, GetTargetsForLevel(logEvent.Level), PrepareLogEventInfo(logEvent), Factory);
+            var targetsForLevel = GetTargetsForLevel(level);
+            if (targetsForLevel != null)
+            {
+                var logEvent = LogEventInfo.Create(level, Name, ex, formatProvider, message, args);
+                WriteToTargets(logEvent, targetsForLevel);
+            }
+        }
+
+        private void WriteToTargets([NotNull] LogEventInfo logEvent, [NotNull] TargetWithFilterChain targetsForLevel)
+        {
+            LoggerImpl.Write(DefaultLoggerType, targetsForLevel, PrepareLogEventInfo(logEvent), Factory);
         }
 
-        private void WriteToTargets(Type wrapperType, LogEventInfo logEvent)
+        private void WriteToTargets(Type wrapperType, [NotNull] LogEventInfo logEvent, [NotNull] TargetWithFilterChain targetsForLevel)
         {
-            LoggerImpl.Write(wrapperType ?? _loggerType, GetTargetsForLevel(logEvent.Level), PrepareLogEventInfo(logEvent), Factory);
+            LoggerImpl.Write(wrapperType ?? DefaultLoggerType, targetsForLevel, PrepareLogEventInfo(logEvent), Factory);
         }
 
         internal void SetConfiguration(LoggerConfiguration newConfiguration)
         {
             _configuration = newConfiguration;
 
             // pre-calculate 'enabled' flags
-            _isTraceEnabled = newConfiguration.IsEnabled(LogLevel.Trace);
-            _isDebugEnabled = newConfiguration.IsEnabled(LogLevel.Debug);
-            _isInfoEnabled = newConfiguration.IsEnabled(LogLevel.Info);
-            _isWarnEnabled = newConfiguration.IsEnabled(LogLevel.Warn);
-            _isErrorEnabled = newConfiguration.IsEnabled(LogLevel.Error);
-            _isFatalEnabled = newConfiguration.IsEnabled(LogLevel.Fatal);
+            _isTraceEnabled = IsEnabled(LogLevel.Trace);
+            _isDebugEnabled = IsEnabled(LogLevel.Debug);
+            _isInfoEnabled = IsEnabled(LogLevel.Info);
+            _isWarnEnabled = IsEnabled(LogLevel.Warn);
+            _isErrorEnabled = IsEnabled(LogLevel.Error);
+            _isFatalEnabled = IsEnabled(LogLevel.Fatal);
 
             OnLoggerReconfigured(EventArgs.Empty);
         }
 
         private TargetWithFilterChain GetTargetsForLevel(LogLevel level)
         {
-            return _configuration.GetTargetsForLevel(level);
+            if (ReferenceEquals(_contextLogger, this))
```

Comment:
this is a performance tweak isn't? Please add a code comment  (before it gets refactored ;))

Reply:
Nope avoid recursive loop.